File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.py
import sys
sys.setrecursionlimit(10 ** 6) # 재귀 횟수 제한 풀기

N, M, R = map(int, sys.stdin.readline().split(" "))

graph = [[] for i in range(N + 1)] # 그래프 생성

ans = [0 for i in range(N + 1)] # 방문 순서 및 방문 여부 동시체크

seq = 1 # 방문 순서 

for i in range(M): # 간선
    a, b = map(int, sys.stdin.readline().split(" "))
    graph[a].append(b)
    graph[b].append(a)

for s in graph: # 그래프 안의 원소들을 내림차순으로 정렬
    s.sort(reverse=True)

def DFS(start):
    global seq
    ans[start] = seq
    for to_v in graph[start]:
        if(ans[to_v]):
            continue
        seq += 1
        DFS(to_v)

DFS(R)

for i in range(1, len(ans)):
    print(ans[i])

File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.java
import java.util.*;
import java.io.*;

public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    private static int[] answer;
    private static int seq = 0;

    private static void DFS(int start) {
        seq += 1;
        answer[start] = seq;

        for(int i = 0 ; i < graph.get(start).size() ;i++) {
            int newVertex = graph.get(start).get(i);
            if(answer[newVertex] == 0) {
                DFS(newVertex);
            }
        }
    }


    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine(), " ");

        int vertex = Integer.parseInt(st.nextToken());
        int edge = Integer.parseInt(st.nextToken());
        int startVertex = Integer.parseInt(st.nextToken());

        answer = new int[vertex + 1];

        for(int i = 0 ; i < vertex + 1 ; i++) {
            graph.add(new ArrayList<>());
        }

        for(int i = 0 ; i < edge ; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int fromVertex = Integer.parseInt(st.nextToken());
            int toVertex = Integer.parseInt(st.nextToken());

            graph.get(fromVertex).add(toVertex);
            graph.get(toVertex).add(fromVertex);
        }

        for(int i = 1 ; i < graph.size() ; i++) {
            Collections.sort(graph.get(i), Collections.reverseOrder());
        }    

        DFS(startVertex);

        for(int i = 1 ; i <= vertex ; i++) {
            bw.write(Integer.toString(answer[i]));
            bw.write("\n");
        }

        bw.flush();

        br.close();
        bw.close();
    }
}